repak/src/pak.rs

```diff
@@ -370,23 +462,179 @@ impl Pak {
             version: self.version,
             version_major: self.version.version_major(),
             index_offset,
-            index_size: index_data.len() as u64,
-            hash: hasher.finalize().into(),
+            index_size: index_buf.len() as u64,
+            hash: index_hash,
             frozen: false,
             compression: vec![],
         };
 
-        writer.write_all(&index_data)?;
-
         footer.write(writer)?;
 
         Ok(())
     }
 }
 
+fn hash(data: &[u8]) -> [u8; 20] {
+    use sha1::{Digest, Sha1};
+    let mut hasher = Sha1::new();
+    hasher.update(data);
+    hasher.finalize().into()
+}
+
+const ENCODED_ENTRY_SIZE: u32 = {
+    4 // flags
+    + 4 // offset
+    + 4 // size
+};
+
+fn generate_path_hash_index<W: Write>(
+    writer: &mut W,
+    path_hash_seed: u64,
+    entries: &BTreeMap<String, super::entry::Entry>,
+) -> Result<(), super::Error> {
+    writer.write_u32::<LE>(entries.len() as u32)?;
+    let mut offset = 0u32;
+    for (path, _) in entries.iter() {
+        let utf16le_path;
+        #[cfg(unix)]
+        {
+            utf16le_path = convert_unix_path_to_utf16le_bytes(path);
+        }
+
+        #[cfg(windows)]
+        {
+            utf16le_path = utf16le_path_to_bytes(p);
+        }
+
+        #[cfg(not(any(unix, windows)))]
+        unimplemented!("unsupported platform");
```

Comment:
Why not do something along these lines to entirely avoid anything platform specific?

        let utf16le_path = path
            .encode_utf16()
            .flat_map(|c| c.to_le_bytes())
            .collect::<Vec<_>>();